### PR TITLE
Fix deterministic group_shuffle_split

### PR DESCRIPTION
**Pull request by nilsleh**
Sets are unordered and therefore, repeated calls were yielding different train and val sizes for Cyclone dataset.

### CODE REVIEW

File: torchgeo/datamodules/utils.py [hunk 1]
```diff
@@ -197,8 +197,9 @@ def group_shuffle_split(
         )
 
     generator = np.random.default_rng(seed=random_state)
+    ordered_list = sorted(list(group_vals))
```
Comment (adamjstewart) [inline torchgeo/datamodules/utils.py:200]: Let's move this up to line 188 and do it all in one go:
```python
group_vals = sorted(list(set(groups)))
```

**Comment by adamjstewart** [comment]
Let's find a way to test this so the same bug doesn't happen again.

**Comment by isaaccorley** [comment]
@nilsleh Would love to get this in the 0.5.2 release!

**Comment by nilsleh** [comment]
> @nilsleh Would love to get this in the 0.5.2 release!

Thanks for the reminder :) Added a unit test to check repeated calls to the function.

**Comment by nilsleh** [comment]
@adamjstewart and @isaaccorley not sure how to simulate repeated calls to the function after restarting script/kernel, so I thought separate processes might be a way to go, but actually not sure

**Comment by adamjstewart** [comment]
> @adamjstewart and @isaaccorley not sure how to simulate repeated calls to the function after restarting script/kernel, so I thought separate processes might be a way to go, but actually not sure

This feels like overkill. Depending on the size of our fake dataset, can we just run the test once, print the order, then hardcode that in the test code? As long as it is always the same, it's deterministic.

File: tests/datamodules/test_utils.py [hunk 2]
```diff
@@ -43,16 +46,21 @@ def test_group_shuffle_split() -> None:
...
+    test_cases = [(None, 0.2, 42), (0.8, None, 42)]
+
+    for train_size, test_size, random_state in test_cases:
+        train_indices1, test_indices1 = group_shuffle_split(
+            groups,
+            train_size=train_size,
+            test_size=test_size,
+            random_state=random_state,
+        )
+        # Check that the results are the same as the saved results
```
Comment (adamjstewart) [inline tests/datamodules/test_utils.py:61]:
```suggestion
        # Check that the results are the same as expected
```
We are no longer "saving" them per se

**Review by adamjstewart — DISMISSED** [review]
Much better, thanks!

File: tests/datamodules/test_utils.py [hunk 3]
```diff
@@ -43,16 +46,21 @@ def test_group_shuffle_split() -> None:
...
+    test_cases = [(None, 0.2, 42), (0.8, None, 42)]
+
+    for train_size, test_size, random_state in test_cases:
+        train_indices1, test_indices1 = group_shuffle_split(
+            groups,
+            train_size=train_size,
+            test_size=test_size,
+            random_state=random_state,
+        )
+        # Check that the results are the same as expcected
```
Comment (adamjstewart) [inline tests/datamodules/test_utils.py:61]:
```suggestion
        # Check that the results are the same as expected
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**